Adds WFS support

tkohr opened the pull request:
This PR enables printing WFS on map using default OL style. Changing opacity is not part of the PR as it belongs to the style.

The spec allows to define the WFS `version` (1.0.0, 1.1.0, 2.0.0) and data `format` (geojson, gml) to use. The GML version used, is currently coupled to the WFS version indicated. (There is still a problem with WFS 2.0.0)

The PR does not monitor the download progress, as all data is loaded at once.

TODOS:

- [x] make WFS 2.0.0 (with GML) work
- [x] make rendering test work with local resource
- [x] add unit test for url generation

fgravin commented on src/printer/layers.js:
```diff
@@ -267,3 +277,120 @@ function createLayerWMTS(layerSpec, rootFrameState) {
+        service: 'WFS',
+        request: 'GetFeature',
+        version: layerSpec.version ? layerSpec.version : '1.1.0',
+        typename: layerSpec.layer,
```
not the same name for WFS 2.0.0

tkohr replied:
Ah that's right, thanks. Somehow WFS 2.0.0 still doesn't work. When I try to open the `GetFeature` url in the browser it downloads the response for 2.0.0. Why? Could this be the reason?

fgravin commented on src/printer/layers.js:
```diff
@@ -267,3 +277,120 @@ function createLayerWMTS(layerSpec, rootFrameState) {
+      if (layerSpec.format !== 'gml') {
+        params['outputFormat'] = 'application/json';
+      }
+      let urlParams = Object.entries(params)
```
Use something like this to build the url https://github.com/camptocamp/mtes-carto-ui/blob/81606d6860dd7dc7f8dcb153c176186307b9dc8d/libs/feature/shared/src/lib/helpers/ogc.helpers.ts#L17

tkohr replied:
Thanks, I adapted

fgravin commented on src/printer/layers.js:
```diff
@@ -267,3 +277,120 @@ function createLayerWMTS(layerSpec, rootFrameState) {
+      let url = layerSpec.url + '?' + urlParams;
+      xhr.open('GET', url);
+      let onError = function () {
+        vectorSource.removeLoadedExtent(extent);
```
Do you raise the error to the client ?

tkohr replied:
Yes, after rebase on merged errorhandling the url is raised into the job status

fgravin commented on src/printer/layers.js:
```diff
@@ -267,3 +277,120 @@ function createLayerWMTS(layerSpec, rootFrameState) {
+    source: vectorSource,
+  });
+
+  frameState = {
```
can this be factorized ?

tkohr replied:
Indeed I have the impression that we only need the `layer` in the `layerStatesArray` for vectorlayers, but I don't think I have understood this part entirely.

fgravin commented on test/unit/printer/layers.test.js:
```diff
@@ -190,6 +190,32 @@ describe('layer creation', () => {
+      );
+    });
+
+    it('initially emit a status with progress 0', () => {
```
can't you test the url of the WFS GetFeatures ?

tkohr replied:
Hum, good point. How do I trigger the `VectorSource.loader()` (to construct the url).

fgravin replied:
I don't know :)

fgravin commented on src/printer/layers.js:
```diff
@@ -267,3 +277,120 @@ function createLayerWMTS(layerSpec, rootFrameState) {
+    this.context = context;
+    this.container = {
+      firstElementChild: context.canvas,
+      style: {
```
is it mandatory as the opacity is in the layer state already ?

tkohr replied:
If I dont' indicate it here I get an error from ol `VectorLayer`

fgravin commented on src/printer/layers.js:
```diff
@@ -267,3 +277,120 @@ function createLayerWMTS(layerSpec, rootFrameState) {
+  };
+
+  renderer = layer.getRenderer();
+  renderer.useContainer = function () {
```
same ?

tkohr replied:
I don't understand

fgravin replied:
Can it be factorized ? Seems it's almost the same for all layers isn't it ?

tkohr replied:
Ah yes of course

fgravin commented on src/printer/layers.js:
```diff
@@ -267,3 +277,120 @@ function createLayerWMTS(layerSpec, rootFrameState) {
+  let vectorSource = new VectorSource({
+    format: format,
+    loader: function (extent, resolution, projection) {
+      let proj = projection.getCode();
```
```suggestion
      const projCode = projection.getCode();
```

fgravin commented on src/printer/layers.js:
```diff
@@ -267,3 +277,120 @@ function createLayerWMTS(layerSpec, rootFrameState) {
+    format: format,
+    loader: function (extent, resolution, projection) {
+      let proj = projection.getCode();
+      let xhr = new XMLHttpRequest();
```
```suggestion
      const xhr = new XMLHttpRequest();
```

fgravin reviewed (commented):
Very good starts, it works well.
Some comments about code optimization.
Tests are a bit light

tkohr commented:
Thanks for your fix @fgravin. Nice that it also works without forcing GML3.

I added a test for url generation and moved rendering test to local resource. Saw CORS issue popping up a couple of times, which disappeared again by itself. From my side PR is ready.
